stage #2.5/module #4 REST Basics - Spring Web/REST Basics/content.md

```diff
@@ -0,0 +1,526 @@
+```
+- ### controller:
+A controller resource models a procedural concept. Controller resources are like executable functions, with parameters and return values, inputs, and outputs.
+_**A REST API relies on controller resources to perform application specific actions that cannot be logically mapped to one of the standard methods (create, retrieve, update and delete).**_
+Use “verb” to denote controller archetype. Controller  “verb” typically appear as the last segment in a URI path, with no child resources to follow them in the hierarchy. 
+```
+http://api.example.com/cart-management/users/{id}/cart/checkout 
+http://api.example.com/song-management/users/{id}/playlist/play
+```
+The example above shows a controller resource that allows a client to checkout his cart or start to play client's playlist.
+
+### Use consistent resource naming conventions and URI formatting
+Using consistent resource naming conventions and URI formatting allows you to minimize ambiguity and maximize readability and maintainability. Below some design hints to achieve consistency:
+- **Use forward slash (/) to indicate hierarchical relationships:**<br>
+The forward-slash (/) character is used in the path portion of the URI to indicate a hierarchical relationship between resources. e.g.
+```
+http://api.example.com/device-management
+http://api.example.com/device-management/managed-devices
+http://api.example.com/device-management/managed-devices/{id}
+http://api.example.com/device-management/managed-devices/{id}/scripts
+http://api.example.com/device-management/managed-devices/{id}/scripts/{id}
+```
+- **Do not use trailing forward slash (/) in URIs:**<br>
+As the last character within a URI’s path, a forward slash (/) adds no semantic value and may confuse. It’s better to drop it from the URI.<br> Examples:<br>
+```
+http://api.example.com/device-management/managed-devices/ 
+http://api.example.com/device-management/managed-devices  /*This is much better version*/
+```
+- **Use hyphens (-) to improve the readability of URIs:**<br>
+To make your URIs easy for people to scan and interpret, use the hyphen (-) character to improve the readability of names in long path segments.<br> Examples:<br>
+```
+http://api.example.com/device-management/managed-devices/
+http://api.example.com/device-management/managed-devices  /*This is much better version*/
+```
+- **Do not use underscores ( _ ):**<br>
+It’s possible to use an underscore in place of a hyphen to be used as a separator – But depending on the application’s font, it is possible that _the underscore (_)_ _character can either get partially obscured or completely hidden in some browsers or screens._
+To avoid this confusion, use hyphens (-) instead of underscores ( _ <br> Examples:<br>
+```
+http://api.example.com/inventory-management/managed-entities/{id}/install-script-location  //More readable
+http://api.example.com/inventory-management/managedEntities/{id}/installScriptLocation  //(_) may be obscured or hidden 
```

Comment:
`http://api.example.com/inventory_management/managed_entities/{id}/install_script_location  //(_) may be obscured or hidden ` - use underscore in example